### PR TITLE
Ssh fix

### LINKED ISSUE #33472: Apiserver is leaking go-routines
We are observing that some clusters are leaking go-routines.

In particular, we observed clusters with 2 or 3 nodes that has 200.000+ go-routines.

However, this seem to be happening only on some clusters (not all of them).
I will try to provide more data when I have it.

@kubernetes/sig-api-machinery

### PR DESCRIPTION
**Pull request by fraenkel**
**What this PR does / why we need it**:
Use the dial timeout provided by the SSH library to prevent the dangling connections.

fixes #33472

**Special notes for your reviewer**:
I did not set the TCP Timeout or Deadlines. That will take more work but wasn't sure if it is needed. More importantly, I wasn't sure what values we would want to use.

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/34302)

### CODE REVIEW

**Comment by brendandburns** [comment]
forwarding to @cjcullen who knows the SSH tunnel stuff.

**Comment by cjcullen** [comment]
Thanks for this!

I think this has 1 of the 2 steps necessary to remove the `timeoutDialer`. The hang we saw in practice was deep inside the `ssh.NewClientConn` function. The config.Timeout value has no effect on that call (see discussion here: https://github.com/golang/go/issues/15113).

We also need to have a timeout-able way to create an `ssh.Conn`. I'm not clear on the best way to do that.

**Comment by cjcullen** [comment]
The `dialer` we use will probably need to look something like this:
```
func (d *realSSHDialer) Dial(network, addr string, config *ssh.ClientConfig) (*ssh.Client, error) {
	conn, err := net.DialTimeout(network, addr, config.Timeout)
	if err != nil {
		return nil, err
	}
	conn.KeepAlive = 30 * time.Second
	c, chans, reqs, err := ssh.NewClientConn(conn, addr, config)
	if err != nil {
		return nil, err
	}
	return ssh.NewClient(c, chans, reqs, nil)
}
```

I think that is what @bradfitz was getting at in https://github.com/kubernetes/kubernetes/issues/33472#issuecomment-252035515

**Comment by fraenkel** [comment]
I get what you are thinking but do realize the keepalive isn't really 30s, but 2 hours+.  That is why I wasn't sure if this is really the fix that will do what we want.
We would have to set more options than just that if we want the window to be on the order of minutes.
Let me see what I can do to reduce the timing. Realize platform matters and its different per platform.

**Comment by fraenkel** [comment]
I used SetReadDeadline which is the only thing that would work well.

**Review by cjcullen — APPROVED** [review]

**Comment by fraenkel** [comment]
Is there anything else we need to tweak here?

**Comment by cjcullen** [comment]
I think this is good to go. LGTM.

File: pkg/ssh/ssh.go [hunk 1]
```diff
@@ -179,20 +189,8 @@ const sshDialTimeout = 150 * time.Second
...
-	go func() {
-		defer runtime.HandleCrash()
-		var err error
-		client, err = d.dialer.Dial(network, addr, config)
-		errCh <- err
-	}()
-	select {
-	case err := <-errCh:
-		return client, err
-	case <-time.After(d.timeout):
```
Comment (bogdando) [inline pkg/ssh/ssh.go]: Wondering why that never interrupted the leaky routine?.. What could we learn from that anti-pattern? "Less (go func() + defer + time.After) == more kittens saved?"

A great solution to the nice bug with corner cases handling and `go func()`!

**Review by bogdando — APPROVED** [review]

File: pkg/ssh/ssh.go (hunk 1 above)
Comment (fraenkel) [inline pkg/ssh/ssh.go]: Nothing we do in the select would actually break the dial.

Comment (bogdando) [inline pkg/ssh/ssh.go]: Then this only confirms the conclusion I made, just never ever use such a pattern as any func calls inside just *can't be trusted*?

**Comment by dims** [comment]
@cjcullen : any concern about the largish "Update golang.org/x/crypto"?

**Comment by dims** [comment]
@saad-ali : In https://github.com/kubernetes/kubernetes/issues/33472 @cjcullen indicated that this is a good PR to merge for 1.5. WDYT?

**Comment by dims** [comment]
So, after doing some more research, i'd like @saad-ali 's input before this merges for 1.5. However if i throw a milestone tag this will get merged immediately, so let me add a do-not-merge and then a v1.5 milestone. So @saad-ali and @cjcullen can clear that tag to get this to merge for 1.5 (let's talk during burndown meeting?)

**Comment by cjcullen** [comment]
@fraenkel, is the crypto godeps update required for the current fix?

**Comment by fraenkel** [comment]
@cjcullen the update is needed to get the dialer timeout code.

**Comment by cjcullen** [comment]
Ahh okay. I'm okay with the dependency update, as long as @saad-ali is.

File: Godeps/Godeps.json [hunk 2]
```diff
@@ -2275,31 +2275,39 @@
...
-			"Rev": "1f22c0103821b9390939b6776727195525381532"
+			"Rev": "d172538b2cfce0c13cee31e647d0367aa8cd2486"
 		},
 		{
 			"ImportPath": "golang.org/x/crypto/curve25519",
-			"Rev": "1f22c0103821b9390939b6776727195525381532"
+			"Rev": "d172538b2cfce0c13cee31e647d0367aa8cd2486"
+		},
+		{
+			"ImportPath": "golang.org/x/crypto/ed25519",
```
Comment (liggitt) [inline Godeps/Godeps.json]: why are we importing new curves and internal packages?

**Comment by saad-ali** [comment]
> Ahh okay. I'm okay with the dependency update, as long as @saad-ali is.

Was the issue this PR is fixing introduced in 1.5? How risky is the fix to the rest of the 1.5 release (modifying the crytpo library is a pretty scary change)? How bad would it be if this PR doesn't make it in to 1.5.0 (i.e. punt it to 1.6.0 or a 1.5.x release)?

**Comment by cjcullen** [comment]
The issue was not introduced in 1.5. The symptom is leaked goroutines when SSH tunnels fail in ssh.Dial (i.e. they got a TCP connection, but TLS handshake was never serviced, or something like that).

If you'd rather punt to 1.6 at this point, I'm fine with it. Apologies to @fraenkel for letting this languish for so long.

**Comment by saad-ali** [comment]
> If you'd rather punt to 1.6 at this point, I'm fine with it. Apologies to @fraenkel for letting this languish for so long.

Thanks @cjcullen. Moving to 1.6 milestone.

File: Godeps/Godeps.json (hunk 2 above)
Comment (fraenkel) [inline Godeps/Godeps.json]: We aren't Godeps pulls in all dependencies and these are new ones for the latest crypto.

**Comment by fraenkel** [comment]
No worries.

**Comment by dims** [comment]
@fraenkel this should get in as soon as master merges open up early next week!

**Comment by rmmh** [comment]
@k8s-bot gce etcd3 e2e test this

**Comment by fraenkel** [comment]
@cjcullen This should be ready to go now that 1.5 is done. I had to rebase so your lgtm got removed.

**Comment by cjcullen** [comment]
Uggh. Sorry I missed this. Rebase one more time and I'll LGTM.

**Comment by fraenkel** [comment]
@cjcullen rebased and passing.

**Pull request merged**